### PR TITLE
feat: log args separately.

### LINKED ISSUE #17542: Cy.log args are no longer printed to the console showing newlines after cypress 7.2
### Current behavior

`cy.log` shows args as an array and doesn't show the newlines that may be present. This make it very difficult to read large debug messages (something my QA users have come to depend on when debugging tests).

### Desired behavior

`cy.log` entries when clicked are printed to the developer console showing newlines. This is helpful if you have large debug logs you wish to view. This is how it used to work in 7.2 and prior. 

### Test code to reproduce

```
"cypress": "7.2.0",  <=== shows newlines in console logs
...
"cypress": "7.3.0",  <=== behavior changed
```

```
it('cy.log works', () => {
  cy.log("Hello there from javascript", "this is not an array\nlarge logs with lots of lines\n\nmore logging\n\neven more")
})
```

### Cypress Version

7.3+

### Other

Cypress 7.2 and earlier:
Screen Shot 2021-07-30 at 3 24 39 PM

Cypress 7.3 and later:
Screen Shot 2021-07-30 at 3 23 45 PM

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #17542 

### User facing changelog

`cy.log` logs its args on console window separately.

### Additional details
- Why was this change necessary? => multiline arg became harder to read from 7.3.
- What is affected by this change? => args are shown in each line.
- Any implementation details to explain? => I wrote a new case for args.

### How has the user experience changed?

**Before:**

image

**After:**

image

### Questions

* I applied this only to `args`, but should we do this for all arrays?
* I couldn't find tests that test the output on the console window. Is there a way to test this change?

### PR Tasks

- [x] Have tests been added/updated?

### CODE REVIEW

File: packages/runner-shared/src/logger.js [hunk 1]
```diff
@@ -54,6 +55,26 @@ export const logger = {
...
+  _logArgs (consoleProps) {
+    const args = this._getArgs(consoleProps)
+
+    if (!args) return
+
+    this.log(`%cArgs:`, 'font-weight: bold')
+
+    args.forEach((arg, index) => {
+      this.log(`%c  [${index}]:`, 'font-weight: bold', arg)
+    })
```
Comment (flotwig) [inline packages/runner-shared/src/logger.js:67]: What do you think about keeping the old behavior if `args.length === 1`, and then only showing the indices if there is more than one element?

**Review by flotwig — CHANGES_REQUESTED** [review]
> I applied this only to args, but should we do this for all arrays?

Are there other places where we pass arrays that this would help?

> I couldn't find tests that test the output on the console window. Is there a way to test this change?

I'm having trouble finding tests for this too. The appropriate place to add tests is `packages/runner/cypress/integration`. But it does need tests.

File: packages/runner-shared/src/logger.js (hunk 1 above)
Comment (sainthkh) [reply packages/runner-shared/src/logger.js:67]: I'll change it that way.

**Comment by sainthkh** [comment]
As the change is helpful for those who want to see the "multiline" values in console window, I guess it doesn't matter to other arguments in array format. 

At that time when I wrote this issue, I thought they might be good for other cases. But I think the Chrome/Firefox default is good enough.

----

I tried mocking `console.log` but it doesn't work. It might be because the code under test uses iframe. 

I think the best possible test might be the unit tests that test mocked `console.log` is called correctly. Should we go this way or is there a good alternative?

**Comment by flotwig** [comment]
> I tried mocking `console.log` but it doesn't work. It might be because the code under test uses iframe.

`cy.window().then((win) => cy.spy(win.console, 'log'))` doesn't work?
 
> I think the best possible test might be the unit tests that test mocked `console.log` is called correctly. Should we go this way or is there a good alternative?

Yeah, that would be fine.

**Comment by sainthkh** [comment]
It happened because of `Attempted to wrap log which is already wrapped` error. I solved it by `restore` the spies before creating my own.

File: packages/runner-shared/src/logger.js (hunk 1 above)
Comment (sainthkh) [reply packages/runner-shared/src/logger.js:67]: After thinking it again, I guess not keeping the old behavior is better. The purpose of this change is to improve the readability of long strings. When we keep the old behavior and the only argument is a long string, it prints like below and it's still hard to read.

```
Args: ["long\n\n\nstring\n\nis here"]
```

With the change in this PR, it's like below:

```
Args:
   [0]: long

string

is here
```

I think the latter is what the user wants and it's easier to read.

**Review by flotwig — APPROVED** [review]

**Pull request merged**